R/test_diseasystore.R

```diff
@@ -0,0 +1,462 @@
+          ds$key_join_features(observable = observable, stratification = NULL, start_date, end_date)
+        )
+      }
+
+
+
+      # Then test combinations with non-NULL stratifications
+      expand.grid(observable     = available_observables,
+                  stratification = available_stratifications) |>
+        purrr::pwalk(~ {
+          # This code may fail (gracefully) in some cases. These we catch here
+          output <- tryCatch({
+            ds$key_join_features(
+              observable = as.character(..1),
+              stratification = eval(parse(text = glue::glue("rlang::quos({..2})"))),
+              start_date, end_date
+            )
+          },
+          error = function(e) {
+            testthat::expect_identical(
+              e$message,
+              paste("(At least one) stratification feature does not match observable aggregator. Not implemented yet.")
+            )
+            return(NULL)
+          },
+          warning = function(w) {
+            checkmate::expect_character(w$message, pattern = "Observable already stratified by")
+            return(NULL)
+          })
+
+          # If the code does not fail, we test the output
+          if (!is.null(output)) {
+            key_join_features_tester(dplyr::collect(output), start_date, end_date)
+          }
+
+          # Stop-gap measure to clear dbplyr_### tables
+          SCDB::get_tables(ds %.% target_conn, show_temp = TRUE) |>
+            dplyr::pull(table) |>
+            purrr::keep(~ stringr::str_detect(., "#?dbplyr_")) |>
+            purrr::walk(~ DBI::dbRemoveTable(ds %.% target_conn, .))
+        })
+
+      rm(ds)
+      invisible(gc())
+    }
+  })
+
+
+  testthat::test_that(glue::glue("{diseasystore_class} key_join fails gracefully"), {
+    testthat::skip_if_not(local)
+
+    for (conn in conn_generator()) {
+
+      # Initialise without start_date and end_date
+      ds <- testthat::expect_no_error(diseasystore_generator$new(verbose = FALSE, target_conn = conn))
+
+
+      # Attempt to perform the possible key_joins
+      available_observables  <- ds$available_features |>
+        purrr::keep(~ startsWith(., "n_") | endsWith(., "_temperature"))
+      available_stratifications <- ds$available_features |>
+        purrr::discard(~ startsWith(., "n_") | endsWith(., "_temperature"))
+
+
+      # Test key_join with malformed inputs
+      expand.grid(observable     = available_observables,
+                  stratification = "non_existent_stratification") |>
+        purrr::pwalk(~ {
+          # This code may fail (gracefully) in some cases. These we catch here
+          output <- tryCatch({
+            ds$key_join_features(
+              observable = as.character(..1),
+              stratification = ..2,
+              start_date = start_date,
+              end_date = end_date
+            )
+          }, error = function(e) {
+            checkmate::expect_character(e$message, pattern = "Must be element of set")
```

Comment:
Maybe the error message should inform about the name of the missing stratification.